{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6834db63-ddf7-4da5-b4ca-b75c0394b149",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('/Users/AngelaLee_1/myenv/lib/python3.12/site-packages')\n",
    "import abc_atlas_access\n",
    "from pathlib import Path\n",
    "from abc_atlas_access.abc_atlas_cache.abc_project_cache import AbcProjectCache\n",
    "download_base = Path('/Users/AngelaLee_1/myenv/lib/python3.12/site-packages')\n",
    "abc_cache = AbcProjectCache.from_s3_cache(download_base)\n",
    "feature_matrix_label = 'WMB-10Xv2-TH'\n",
    "file = abc_cache.get_data_path(directory='WMB-10Xv2', file_name='WMB-10Xv2-TH/log2')\n",
    "\n",
    "cell = abc_cache.get_metadata_dataframe(\n",
    "    directory='WMB-10X',\n",
    "    file_name='cell_metadata',\n",
    "    dtype={'cell_label': str}\n",
    ")\n",
    "cell.set_index('cell_label', inplace=True)\n",
    "cluster_details = abc_cache.get_metadata_dataframe(\n",
    "    directory='WMB-taxonomy',\n",
    "    file_name='cluster_to_cluster_annotation_membership_pivoted',\n",
    "    keep_default_na=False\n",
    ")\n",
    "cluster_details.set_index('cluster_alias', inplace=True)\n",
    "cell_extended = cell.join(cluster_details, on='cluster_alias')\n",
    "\n",
    "import anndata\n",
    "adata = anndata.read_h5ad(file, backed='r')\n",
    "gene_filtered = adata.var.iloc[:200, :] #Change the gene_filtered range \n",
    "asubset = adata[:, gene_filtered.index].to_memory()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fd9753d8-0972-4fe2-8de5-5e4671f7c27b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Xkr4</th>\n",
       "      <th>Gm1992</th>\n",
       "      <th>Gm19938</th>\n",
       "      <th>Gm37381</th>\n",
       "      <th>Rp1</th>\n",
       "      <th>Sox17</th>\n",
       "      <th>Gm37587</th>\n",
       "      <th>Gm37323</th>\n",
       "      <th>Mrpl15</th>\n",
       "      <th>Lypla1</th>\n",
       "      <th>...</th>\n",
       "      <th>Sema4c</th>\n",
       "      <th>D430040D24Rik</th>\n",
       "      <th>Fam178b</th>\n",
       "      <th>Cox5b</th>\n",
       "      <th>Actr1b</th>\n",
       "      <th>4933424G06Rik</th>\n",
       "      <th>Gm33533</th>\n",
       "      <th>Zap70</th>\n",
       "      <th>Tmem131</th>\n",
       "      <th>Gm38115</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_label</th>\n",
       "      <th>cluster_alias</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAACGAG-103_A01</th>\n",
       "      <th>5230</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.109224</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAACGCC-042_A01</th>\n",
       "      <th>5033</th>\n",
       "      <td>9.720479</td>\n",
       "      <td>6.732400</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.306012</td>\n",
       "      <td>6.732400</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAAGTGG-040_B01</th>\n",
       "      <th>5023</th>\n",
       "      <td>7.767653</td>\n",
       "      <td>6.774258</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.221658</td>\n",
       "      <td>6.774258</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.774258</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAAGTGG-044_D01</th>\n",
       "      <th>5046</th>\n",
       "      <td>8.461960</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.46196</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.466046</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACAATC-036_D01</th>\n",
       "      <th>5094</th>\n",
       "      <td>7.186204</td>\n",
       "      <td>8.181243</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.764547</td>\n",
       "      <td>7.186204</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.181243</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 204 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            Xkr4    Gm1992  Gm19938  Gm37381  \\\n",
       "cell_label               cluster_alias                                         \n",
       "AAACCTGAGAAACGAG-103_A01 5230           0.000000  0.000000  0.00000      0.0   \n",
       "AAACCTGAGAAACGCC-042_A01 5033           9.720479  6.732400  0.00000      0.0   \n",
       "AAACCTGAGAAAGTGG-040_B01 5023           7.767653  6.774258  0.00000      0.0   \n",
       "AAACCTGAGAAAGTGG-044_D01 5046           8.461960  0.000000  8.46196      0.0   \n",
       "AAACCTGAGAACAATC-036_D01 5094           7.186204  8.181243  0.00000      0.0   \n",
       "\n",
       "                                        Rp1  Sox17  Gm37587  Gm37323  \\\n",
       "cell_label               cluster_alias                                 \n",
       "AAACCTGAGAAACGAG-103_A01 5230           0.0    0.0      0.0      0.0   \n",
       "AAACCTGAGAAACGCC-042_A01 5033           0.0    0.0      0.0      0.0   \n",
       "AAACCTGAGAAAGTGG-040_B01 5023           0.0    0.0      0.0      0.0   \n",
       "AAACCTGAGAAAGTGG-044_D01 5046           0.0    0.0      0.0      0.0   \n",
       "AAACCTGAGAACAATC-036_D01 5094           0.0    0.0      0.0      0.0   \n",
       "\n",
       "                                          Mrpl15    Lypla1  ...  Sema4c  \\\n",
       "cell_label               cluster_alias                      ...           \n",
       "AAACCTGAGAAACGAG-103_A01 5230           0.000000  0.000000  ...     0.0   \n",
       "AAACCTGAGAAACGCC-042_A01 5033           0.000000  0.000000  ...     0.0   \n",
       "AAACCTGAGAAAGTGG-040_B01 5023           0.000000  0.000000  ...     0.0   \n",
       "AAACCTGAGAAAGTGG-044_D01 5046           0.000000  0.000000  ...     0.0   \n",
       "AAACCTGAGAACAATC-036_D01 5094           8.764547  7.186204  ...     0.0   \n",
       "\n",
       "                                        D430040D24Rik  Fam178b      Cox5b  \\\n",
       "cell_label               cluster_alias                                      \n",
       "AAACCTGAGAAACGAG-103_A01 5230                     0.0      0.0   9.109224   \n",
       "AAACCTGAGAAACGCC-042_A01 5033                     0.0      0.0   9.306012   \n",
       "AAACCTGAGAAAGTGG-040_B01 5023                     0.0      0.0  10.221658   \n",
       "AAACCTGAGAAAGTGG-044_D01 5046                     0.0      0.0   7.466046   \n",
       "AAACCTGAGAACAATC-036_D01 5094                     0.0      0.0   8.181243   \n",
       "\n",
       "                                          Actr1b  4933424G06Rik  Gm33533  \\\n",
       "cell_label               cluster_alias                                     \n",
       "AAACCTGAGAAACGAG-103_A01 5230           0.000000            0.0      0.0   \n",
       "AAACCTGAGAAACGCC-042_A01 5033           6.732400            0.0      0.0   \n",
       "AAACCTGAGAAAGTGG-040_B01 5023           6.774258            0.0      0.0   \n",
       "AAACCTGAGAAAGTGG-044_D01 5046           0.000000            0.0      0.0   \n",
       "AAACCTGAGAACAATC-036_D01 5094           0.000000            0.0      0.0   \n",
       "\n",
       "                                        Zap70   Tmem131  Gm38115  \n",
       "cell_label               cluster_alias                            \n",
       "AAACCTGAGAAACGAG-103_A01 5230             0.0  0.000000      0.0  \n",
       "AAACCTGAGAAACGCC-042_A01 5033             0.0  0.000000      0.0  \n",
       "AAACCTGAGAAAGTGG-040_B01 5023             0.0  6.774258      0.0  \n",
       "AAACCTGAGAAAGTGG-044_D01 5046             0.0  0.000000      0.0  \n",
       "AAACCTGAGAACAATC-036_D01 5094             0.0  0.000000      0.0  \n",
       "\n",
       "[5 rows x 204 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Gene expression data per cell\n",
    "pred = (cell_extended['feature_matrix_label'] == feature_matrix_label)\n",
    "cell_filtered = cell_extended[pred]\n",
    "cell_filtered.head()\n",
    "\n",
    "def create_expression_dataframe(ad, gf) :\n",
    "    gdata = ad[:, gf.index].to_df()\n",
    "    gdata.columns = gf.gene_symbol\n",
    "    joined = cell_filtered.join(gdata)\n",
    "    return joined\n",
    "\n",
    "def aggregate_by_metadata(df,gene_filtered, values, sort=False):\n",
    "    grouped = df.groupby(values)[gene_filtered].mean()\n",
    "    if sort:\n",
    "        grouped = grouped.sort_values(by=gene_filtered[0], ascending=False)\n",
    "    return grouped\n",
    "ntexp = create_expression_dataframe(asubset, gene_filtered)\n",
    "agg = aggregate_by_metadata(ntexp, gene_filtered.gene_symbol, values=['cell_label','cluster_alias'])\n",
    "agg.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a7bda5a8-8af4-42dd-b89e-0155ab682eb8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   Xkr4    Gm1992   Gm19938  Gm37381       Rp1  Sox17  \\\n",
      "cluster_alias                                                           \n",
      "1745           7.947417  0.983436  2.049377      0.0  0.000000    0.0   \n",
      "1790           8.948236  1.184351  3.435622      0.0  0.038038    0.0   \n",
      "1794           9.588399  1.500440  4.507796      0.0  0.031238    0.0   \n",
      "2200           8.257760  1.354365  3.131485      0.0  0.000000    0.0   \n",
      "2301           7.639717  2.016544  2.172362      0.0  2.172362    0.0   \n",
      "...                 ...       ...       ...      ...       ...    ...   \n",
      "14983          0.000000  0.000000  0.000000      0.0  0.000000    0.0   \n",
      "24467          7.884557  0.788447  2.662279      0.0  0.000000    0.0   \n",
      "24472          7.735545  0.622895  1.843965      0.0  0.000000    0.0   \n",
      "24473          8.582306  0.645123  2.987764      0.0  0.000000    0.0   \n",
      "24574          8.214292  1.147497  3.189295      0.0  0.000000    0.0   \n",
      "\n",
      "               Gm37587  Gm37323    Mrpl15    Lypla1  ...    Sema4c  \\\n",
      "cluster_alias                                        ...             \n",
      "1745               0.0      0.0  0.796816  0.980263  ...  0.362427   \n",
      "1790               0.0      0.0  1.497515  2.771239  ...  0.149495   \n",
      "1794               0.0      0.0  1.589636  2.880888  ...  0.235358   \n",
      "2200               0.0      0.0  1.645437  2.003980  ...  0.423957   \n",
      "2301               0.0      0.0  0.000000  2.157977  ...  0.000000   \n",
      "...                ...      ...       ...       ...  ...       ...   \n",
      "14983              0.0      0.0  0.000000  0.000000  ...  0.000000   \n",
      "24467              0.0      0.0  0.950255  1.607768  ...  0.000000   \n",
      "24472              0.0      0.0  1.256401  1.351273  ...  0.417074   \n",
      "24473              0.0      0.0  1.747955  2.304741  ...  0.074994   \n",
      "24574              0.0      0.0  0.202300  1.758914  ...  0.364339   \n",
      "\n",
      "               D430040D24Rik  Fam178b     Cox5b    Actr1b  4933424G06Rik  \\\n",
      "cluster_alias                                                              \n",
      "1745                0.066710      0.0  6.033353  2.760019       0.445044   \n",
      "1790                0.148889      0.0  8.517682  5.928851       0.142303   \n",
      "1794                0.103234      0.0  8.918152  5.493396       0.083841   \n",
      "2200                0.203677      0.0  7.559628  4.011184       1.314261   \n",
      "2301                0.000000      0.0  8.195742  4.330339       2.157977   \n",
      "...                      ...      ...       ...       ...            ...   \n",
      "14983               0.000000      0.0  0.000000  0.000000       0.000000   \n",
      "24467               0.275593      0.0  6.680530  4.705503       0.444579   \n",
      "24472               0.089662      0.0  7.777498  4.507771       0.421269   \n",
      "24473               0.000000      0.0  8.175779  5.002680       0.463658   \n",
      "24574               0.000000      0.0  7.657552  3.704976       0.369327   \n",
      "\n",
      "                Gm33533     Zap70   Tmem131   Gm38115  \n",
      "cluster_alias                                          \n",
      "1745           0.296275  0.129792  1.950603  0.069773  \n",
      "1790           0.035684  0.034485  2.879013  0.217328  \n",
      "1794           0.000000  0.085105  3.678660  0.081446  \n",
      "2200           0.299950  0.312982  2.526691  0.095255  \n",
      "2301           0.000000  4.330339  5.180798  0.000000  \n",
      "...                 ...       ...       ...       ...  \n",
      "14983          0.000000  0.000000  0.000000  0.000000  \n",
      "24467          0.509134  0.135252  1.867504  0.247707  \n",
      "24472          0.315495  0.498468  3.044846  0.187026  \n",
      "24473          0.305124  0.298270  2.500964  0.080959  \n",
      "24574          0.000000  0.162796  3.232081  0.000000  \n",
      "\n",
      "[123 rows x 204 columns]\n"
     ]
    }
   ],
   "source": [
    "#Finding the mean gene expression\n",
    "# Assuming you have already defined `cell_filtered`, `gene_filtered`, and `asubset` as per your previous context\n",
    "\n",
    "# Define function to create expression dataframe\n",
    "def create_expression_dataframe(ad, gf):\n",
    "    gdata = ad[:, gf.index].to_df()\n",
    "    gdata.columns = gf.gene_symbol\n",
    "    joined = cell_filtered.join(gdata)\n",
    "    return joined\n",
    "\n",
    "# Define function to aggregate by metadata\n",
    "def aggregate_by_metadata(df, gene_symbols):\n",
    "    grouped = df.groupby('cluster_alias')[gene_symbols].mean()\n",
    "    return grouped\n",
    "\n",
    "# Create expression dataframe\n",
    "ntexp = create_expression_dataframe(asubset, gene_filtered)\n",
    "agg_stats = aggregate_by_metadata(ntexp, gene_filtered.gene_symbol)\n",
    "\n",
    "selected_cluster_aliases = [1745, 1790, 1794, 2200, 2301, 2303, 2304, 2309, 2310, 2444, 2879, 2881, 2884, 2887, 2895, 2898, 2899, 2900, 2901, 2902, 2903, 2904, 2905, 2906, 2907, 2908, 2910, 2911, 2912, 2913, 3425, 3432, 3434, 3490, 4486, 4517,\n",
    "                    4687, 4690, 4702, 4708, 4709, 4721, 4728, 4729, 4730, 4731, 4732, 4735, 4737, 4770, 4837, 4839, 4840, 4844, 4846, 4851, 4852, 4856, 4858, 4861, 5031, 5035, 5036, 5038, 5039, 5044, 5048, 5049, \n",
    "                    5050, 5054, 5057, 5058, 5062, 5077, 5081, 5085, 5090, 5094, 5097, 5101, 5106, 5112, 5113, 5137, 5139, 5203, 5204, 5211, 5212, 5213, 5214, 5215, 5218, 5230, 5231, 5232, 5233, 5234, 5236, 5237, 5243, 5244, 5249, 5250, 5252, 5254, 5255, \n",
    "                    5261, 5262, 5268, 5274, 5283, 14865, 14930, 14956, 14962, 14977, 14980, 14983, 24467, 24472, 24473, 24574]  # Replace with your desired clusters\n",
    "len(selected_cluster_aliases)\n",
    "# Filter agg_stats to include only selected cluster_alias values\n",
    "filtered_agg_stats = agg_stats.loc[selected_cluster_aliases]\n",
    "\n",
    "# Print or display the filtered aggregated data\n",
    "print(filtered_agg_stats)\n",
    "filtered_agg_stats.to_csv('0-200_mean_v2_TH.csv', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1134a1ef-e1a9-4b40-bc27-a3caca856e87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   Xkr4    Gm1992   Gm19938  Gm37381       Rp1  Sox17  \\\n",
      "cluster_alias                                                           \n",
      "1745           2.609710  2.562854  3.405516      0.0  0.000000    0.0   \n",
      "1790           1.627426  2.582834  3.601335      0.0  0.504630    0.0   \n",
      "1794           0.994686  2.810119  3.476700      0.0  0.492929    0.0   \n",
      "2200           2.034112  2.631150  3.456773      0.0  0.000000    0.0   \n",
      "2301           0.984192  3.492757  3.762641      0.0  3.762641    0.0   \n",
      "...                 ...       ...       ...      ...       ...    ...   \n",
      "14983               NaN       NaN       NaN      NaN       NaN    NaN   \n",
      "24467          2.425629  2.209371  3.562544      0.0  0.000000    0.0   \n",
      "24472          2.274024  2.011945  3.090787      0.0  0.000000    0.0   \n",
      "24473          1.181611  2.028023  3.446927      0.0  0.000000    0.0   \n",
      "24574          2.617324  2.645905  3.528707      0.0  0.000000    0.0   \n",
      "\n",
      "               Gm37587  Gm37323    Mrpl15    Lypla1  ...    Sema4c  \\\n",
      "cluster_alias                                        ...             \n",
      "1745               0.0      0.0  2.285674  2.553752  ...  1.598240   \n",
      "1790               0.0      0.0  2.785847  3.450993  ...  0.984944   \n",
      "1794               0.0      0.0  2.814255  3.410936  ...  1.219658   \n",
      "2200               0.0      0.0  2.920619  3.140840  ...  1.618491   \n",
      "2301               0.0      0.0  0.000000  3.737726  ...  0.000000   \n",
      "...                ...      ...       ...       ...  ...       ...   \n",
      "14983              NaN      NaN       NaN       NaN  ...       NaN   \n",
      "24467              0.0      0.0  2.437048  2.971737  ...  0.000000   \n",
      "24472              0.0      0.0  2.668298  2.746688  ...  1.640341   \n",
      "24473              0.0      0.0  2.987048  3.250145  ...  0.691408   \n",
      "24574              0.0      0.0  1.230541  3.151609  ...  1.545988   \n",
      "\n",
      "               D430040D24Rik  Fam178b     Cox5b    Actr1b  4933424G06Rik  \\\n",
      "cluster_alias                                                              \n",
      "1745                0.667098      0.0  3.754994  3.721282       1.772588   \n",
      "1790                0.982023      0.0  1.534966  3.198752       0.936718   \n",
      "1794                0.810216      0.0  1.473926  3.249872       0.761697   \n",
      "2200                1.117839      0.0  2.638739  3.633606       2.691438   \n",
      "2301                0.000000      0.0  0.599578  3.750246       3.737726   \n",
      "...                      ...      ...       ...       ...            ...   \n",
      "14983                    NaN      NaN       NaN       NaN            NaN   \n",
      "24467               1.391675      0.0  3.154484  3.651933       1.817882   \n",
      "24472               0.766635      0.0  2.161266  3.539069       1.657313   \n",
      "24473               0.000000      0.0  1.498182  3.389759       1.695567   \n",
      "24574               0.000000      0.0  2.432209  3.894350       1.566721   \n",
      "\n",
      "                Gm33533     Zap70   Tmem131   Gm38115  \n",
      "cluster_alias                                          \n",
      "1745           1.467167  0.918279  3.321378  0.697728  \n",
      "1790           0.473401  0.457493  3.378766  1.161716  \n",
      "1794           0.000000  0.776169  3.450219  0.740981  \n",
      "2200           1.330091  1.388430  3.298704  0.743963  \n",
      "2301           0.000000  3.750246  4.494346  0.000000  \n",
      "...                 ...       ...       ...       ...  \n",
      "14983               NaN       NaN       NaN       NaN  \n",
      "24467          1.783777  0.975317  3.118310  1.250847  \n",
      "24472          1.417828  1.759364  3.498263  1.122448  \n",
      "24473          1.382359  1.351077  3.250218  0.746401  \n",
      "24574          0.000000  0.990251  3.574008  0.000000  \n",
      "\n",
      "[123 rows x 204 columns]\n"
     ]
    }
   ],
   "source": [
    "#Finding the standard deviation\n",
    "# Assuming you have already defined `cell_filtered`, `gene_filtered`, and `asubset` as per your previous context\n",
    "\n",
    "# Define function to create expression dataframe\n",
    "def create_expression_dataframe(ad, gf):\n",
    "    gdata = ad[:, gf.index].to_df()\n",
    "    gdata.columns = gf.gene_symbol\n",
    "    joined = cell_filtered.join(gdata)\n",
    "    return joined\n",
    "\n",
    "# Define function to aggregate by metadata using standard deviation\n",
    "def aggregate_by_metadata_std(df, gene_symbols):\n",
    "    grouped_std = df.groupby('cluster_alias')[gene_symbols].std()\n",
    "    return grouped_std\n",
    "\n",
    "# Create expression dataframe\n",
    "ntexp = create_expression_dataframe(asubset, gene_filtered)\n",
    "\n",
    "# Aggregate by metadata to get standard deviation\n",
    "agg_stats_std = aggregate_by_metadata_std(ntexp, gene_filtered.gene_symbol)\n",
    "\n",
    "selected_cluster_aliases = [1745, 1790, 1794, 2200, 2301, 2303, 2304, 2309, 2310, 2444, 2879, 2881, 2884, 2887, 2895, 2898, 2899, 2900, 2901, 2902, 2903, 2904, 2905, 2906, 2907, 2908, 2910, 2911, 2912, 2913, 3425, 3432, 3434, 3490, 4486, 4517,\n",
    "                    4687, 4690, 4702, 4708, 4709, 4721, 4728, 4729, 4730, 4731, 4732, 4735, 4737, 4770, 4837, 4839, 4840, 4844, 4846, 4851, 4852, 4856, 4858, 4861, 5031, 5035, 5036, 5038, 5039, 5044, 5048, 5049, \n",
    "                    5050, 5054, 5057, 5058, 5062, 5077, 5081, 5085, 5090, 5094, 5097, 5101, 5106, 5112, 5113, 5137, 5139, 5203, 5204, 5211, 5212, 5213, 5214, 5215, 5218, 5230, 5231, 5232, 5233, 5234, 5236, 5237, 5243, 5244, 5249, 5250, 5252, 5254, 5255, \n",
    "                    5261, 5262, 5268, 5274, 5283, 14865, 14930, 14956, 14962, 14977, 14980, 14983, 24467, 24472, 24473, 24574]  # Replace with your desired clusters\n",
    "\n",
    "# Filter agg_stats_std to include only selected cluster_alias values\n",
    "filtered_agg_stats_std = agg_stats_std.loc[selected_cluster_aliases]\n",
    "\n",
    "# Print or display the filtered aggregated data\n",
    "print(filtered_agg_stats_std)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "0fcc8aac-3537-46ff-b0e7-510a66a3c4d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        cluster_alias           gene  cell_count\n",
      "27800            1745  0610009E02Rik         100\n",
      "27801            1745  1700001O22Rik         100\n",
      "27802            1745  1700007K13Rik         100\n",
      "27803            1745  1700084E18Rik         100\n",
      "27804            1745  1810012K08Rik         100\n",
      "...               ...            ...         ...\n",
      "186995          24574           Vav2          37\n",
      "186996          24574          Wdr34          37\n",
      "186997          24574           Wdr5          37\n",
      "186998          24574        Zdhhc12          37\n",
      "186999          24574           Zer1          37\n",
      "\n",
      "[24600 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "#Finding the n-value\n",
    "# Assuming you have already defined `cell_filtered`, `gene_filtered`, and `asubset` as per your previous context\n",
    "\n",
    "# Define function to create expression dataframe\n",
    "def create_expression_dataframe(ad, gf):\n",
    "    gdata = ad[:, gf.index].to_df()\n",
    "    gdata.columns = gf.gene_symbol\n",
    "    joined = cell_filtered.join(gdata)\n",
    "    return joined\n",
    "\n",
    "# Define function to count number of cells in each cluster_alias for each gene\n",
    "def count_cells_per_gene(df, gene_symbols):\n",
    "    # Melting the dataframe to have a long format suitable for counting\n",
    "    melted_df = df.melt(id_vars=['cluster_alias'], value_vars=gene_symbols, var_name='gene', value_name='expression')\n",
    "    # Filtering out rows where expression is NaN\n",
    "    melted_df = melted_df.dropna(subset=['expression'])\n",
    "    # Counting the number of cells for each gene in each cluster_alias\n",
    "    count_df = melted_df.groupby(['cluster_alias', 'gene']).size().reset_index(name='cell_count')\n",
    "    return count_df\n",
    "\n",
    "# Create expression dataframe\n",
    "ntexp = create_expression_dataframe(asubset, gene_filtered)\n",
    "\n",
    "# Get list of gene symbols\n",
    "gene_symbols = gene_filtered.gene_symbol.tolist()\n",
    "\n",
    "# Count number of cells in each cluster_alias for each gene\n",
    "cell_counts = count_cells_per_gene(ntexp, gene_symbols)\n",
    "\n",
    "# Filter cell counts to include only selected cluster_alias values\n",
    "selected_cluster_aliases = [1745, 1790, 1794, 2200, 2301, 2303, 2304, 2309, 2310, 2444, 2879, 2881, 2884, 2887, 2895, 2898, 2899, 2900, 2901, 2902, 2903, 2904, 2905, 2906, 2907, 2908, 2910, 2911, 2912, 2913, 3425, 3432, 3434, 3490, 4486, 4517,\n",
    "                    4687, 4690, 4702, 4708, 4709, 4721, 4728, 4729, 4730, 4731, 4732, 4735, 4737, 4770, 4837, 4839, 4840, 4844, 4846, 4851, 4852, 4856, 4858, 4861, 5031, 5035, 5036, 5038, 5039, 5044, 5048, 5049, \n",
    "                    5050, 5054, 5057, 5058, 5062, 5077, 5081, 5085, 5090, 5094, 5097, 5101, 5106, 5112, 5113, 5137, 5139, 5203, 5204, 5211, 5212, 5213, 5214, 5215, 5218, 5230, 5231, 5232, 5233, 5234, 5236, 5237, 5243, 5244, 5249, 5250, 5252, 5254, 5255, \n",
    "                    5261, 5262, 5268, 5274, 5283, 14865, 14930, 14956, 14962, 14977, 14980, 14983, 24467, 24472, 24473, 24574]  # Replace with your desired clusters\n",
    "\n",
    "filtered_cell_counts = cell_counts[cell_counts['cluster_alias'].isin(selected_cluster_aliases)]\n",
    "\n",
    "# Print or display the filtered cell counts\n",
    "print(filtered_cell_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "c490fedc-d6cf-4728-8297-94913b63471e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene Gm1992 not found in file 201-400_mean_v2_TH.csv.\n",
      "                 Filename  cluster_alias_original    Gm1992  cluster_alias\n",
      "0    0-200_mean_v2_TH.csv                    1745  0.983436           1745\n",
      "1    0-200_mean_v2_TH.csv                    1790  1.184351           1790\n",
      "2    0-200_mean_v2_TH.csv                    1794  1.500440           1794\n",
      "3    0-200_mean_v2_TH.csv                    2200  1.354365           2200\n",
      "4    0-200_mean_v2_TH.csv                    2301  2.016544           2301\n",
      "..                    ...                     ...       ...            ...\n",
      "118  0-200_mean_v2_TH.csv                   14983  0.000000          14983\n",
      "119  0-200_mean_v2_TH.csv                   24467  0.788447          24467\n",
      "120  0-200_mean_v2_TH.csv                   24472  0.622896          24472\n",
      "121  0-200_mean_v2_TH.csv                   24473  0.645123          24473\n",
      "122  0-200_mean_v2_TH.csv                   24574  1.147497          24574\n",
      "\n",
      "[123 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "def find_gene_in_files(directory, gene):\n",
    "    gene_data = []\n",
    "    gene_files = []\n",
    "    \n",
    "    # Iterate through files in the directory\n",
    "    for filename in os.listdir(directory):\n",
    "        if filename.endswith('.csv'):  # Adjust if your files have a different extension\n",
    "            file_path = os.path.join(directory, filename)\n",
    "            \n",
    "            # Read the file\n",
    "            df = pd.read_csv(file_path)\n",
    "            \n",
    "            # Check if the gene is in the columns\n",
    "            if gene in df.columns:\n",
    "                # Extract the data for the specific gene and the cluster_alias\n",
    "                gene_values = df[[gene, 'cluster_alias']]\n",
    "                \n",
    "                # Add the original cluster_alias column from the file as the first column\n",
    "                gene_values.insert(0, 'cluster_alias_original', df['cluster_alias'])\n",
    "                \n",
    "                gene_data.append(gene_values)\n",
    "                gene_files.append(filename)\n",
    "            else:\n",
    "                print(f\"Gene {gene} not found in file {filename}.\")\n",
    "    \n",
    "    # Combine data from multiple files\n",
    "    if gene_data:\n",
    "        combined_data = pd.concat(gene_data, keys=gene_files, names=['Filename'])\n",
    "        combined_data = combined_data.reset_index(level='Filename')\n",
    "        return combined_data\n",
    "    else:\n",
    "        print(f\"Gene {gene} not found in any files.\")\n",
    "        return None\n",
    "\n",
    "# Example usage\n",
    "directory = '/Users/AngelaLee_1/Downloads/Fos-Seq/10X TH Gene Mean'  # Replace with the path to your directory containing the CSV files\n",
    "gene = 'Gm1992'  # Replace with the gene you are interested in\n",
    "\n",
    "combined_gene_data = find_gene_in_files(directory, gene)\n",
    "\n",
    "if combined_gene_data is not None:\n",
    "    # Print or display the combined gene data\n",
    "    print(combined_gene_data)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
